File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/my_number_counter/scripts/number_counter.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import Int64
from std_srvs.srv import SetBool

counter = 0
pub =None

def callback_receive_number_data(msg):
    global counter
    rospy.loginfo("Message Received")
    rospy.loginfo(msg)#msg.data also fine
    counter += msg.data
    new_msg = Int64()
    new_msg.data = counter
    pub.publish(new_msg)

def reset_count(req):
    if req.data:
        global counter
        rospy.loginfo("Counter Reset!")
        counter = 0
        return True, "Counter Reset Success"

    else:
        return False, "Counter Reset False"

if __name__ == '__main__':
    rospy.init_node('number_counter')
    
    #subscriber
    sub = rospy.Subscriber("/number", Int64, callback_receive_number_data)
    
    #publisher
    pub = rospy.Publisher("/number_count", Int64, queue_size=10)

    #create Service server
    service = rospy.Service("/reset_number_count", SetBool, reset_count)

    rospy.spin()

File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/my_number_counter/src/number_counter.cpp
#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <std_srvs/SetBool.h>

int counter = 0;
ros::Publisher pub;

void callback_receive_number_data(const std_msgs::Int64& msg)
{
	ROS_INFO("Message Received : %li", msg.data);
	counter += msg.data;
	std_msgs::Int64 new_msg;
	new_msg.data = counter;
	pub.publish(new_msg);
}

bool reset_number_count(std_srvs::SetBool::Request &req, std_srvs::SetBool::Response &res){
	if (req.data){
		ROS_INFO("Reset Succeed");
		counter = 0;
		res.success = true;
		res.message = "Counter Reset Success";
	}
	else{
		res.success = false;
		res.message = "Counter Reset False";
	}
	return true;
}

int main(int argc, char **argv)
{
	ros::init(argc, argv, "number_counter");

	ros::NodeHandle nh;

	ros::Subscriber sub = nh.subscribe("/number", 1000, callback_receive_number_data);

	pub = nh.advertise<std_msgs::Int64>("/number_count", 10);
	
	// Create Server
	ros::ServiceServer server = nh.advertiseService("/reset_number_count", reset_number_count);

	ros::spin(); //Keep node running till we shutdown
}
